internal/pkg/peer/blocksprovider/bft_header_receiver.go

```diff
@@ -0,0 +1,164 @@
+)
+
+const bftHeaderWrongStatusThreshold = 10
+
+type bftHeaderReceiver struct {
+	mutex              sync.Mutex
+	chainID            string
+	stop               bool
+	stopChan           chan struct{}
+	started            bool
+	endpoint           string
+	client             orderer.AtomicBroadcast_DeliverClient
+	msgCryptoVerifier  BlockVerifier
+	lastHeader         *common.Block // a block with Header & Metadata, without Data (i.e. lastHeader.Data==nil)
+	lastHeaderTime     time.Time
+	lastHeaderVerified bool
+	lastHeaderOK       bool
+	logger             *flogging.FabricLogger
+}
+
+func NewBFTHeaderReceiver(
+	chainID string,
+	endpoint string,
+	client orderer.AtomicBroadcast_DeliverClient,
+	msgVerifier BlockVerifier,
+	logger *flogging.FabricLogger,
+) *bftHeaderReceiver {
+	hRcv := &bftHeaderReceiver{
+		chainID:           chainID,
+		stopChan:          make(chan struct{}, 1),
+		endpoint:          endpoint,
+		client:            client,
+		msgCryptoVerifier: msgVerifier,
+		logger:            logger.With("endpoint", endpoint),
+	}
+	return hRcv
+}
+
+// DeliverHeaders starts to deliver headers from the stream client
+func (hr *bftHeaderReceiver) DeliverHeaders() {
+	defer func() {
+		_ = hr.CloseSend()
+	}()
+
+	hr.logger.Debugf("[%s] Starting to deliver headers", hr.chainID)
+	hr.setStarted()
+
+	for !hr.IsStopped() {
+		msg, err := hr.client.Recv()
+		if err != nil {
+			hr.logger.Warningf("[%s] Receive error: %s", hr.chainID, err.Error())
+			return
+		}
+
+		switch t := msg.GetType().(type) {
+		case *orderer.DeliverResponse_Status:
+			if t.Status == common.Status_SUCCESS {
+				hr.logger.Warningf("[%s] Warning! Received %s for a seek that should never complete", hr.chainID, t.Status)
+				return
+			}
+
+			hr.logger.Errorf("[%s] Got bad status %s", hr.chainID, t.Status)
+			return
+
+		case *orderer.DeliverResponse_Block:
+			blockNum := t.Block.Header.Number
+
+			// do not verify, just save for later, in case the block-receiver is suspected of censorship
+			hr.logger.Debugf("[%s][%s] Saving block with header & metadata, blockNum = [%d], block = [%v]", hr.chainID, hr.endpoint, blockNum, t.Block)
+			hr.mutex.Lock()
+			hr.lastHeader = t.Block
+			hr.lastHeaderTime = time.Now()
+			hr.lastHeaderVerified = false
+			hr.lastHeaderOK = false
+			hr.mutex.Unlock()
+
+		default:
+			hr.logger.Warningf("[%s] Received unknown response type: %v", hr.chainID, t)
+			return
+		}
+	}
+
+	hr.logger.Debugf("[%s] Stopped to deliver headers from endpoint: %s", hr.chainID, hr.endpoint)
+}
+
+func (hr *bftHeaderReceiver) IsStopped() bool {
+	hr.mutex.Lock()
+	defer hr.mutex.Unlock()
+	return hr.stop
+}
+
+func (hr *bftHeaderReceiver) IsStarted() bool {
+	hr.mutex.Lock()
+	defer hr.mutex.Unlock()
+	return hr.started
+}
+
+func (hr *bftHeaderReceiver) setStarted() {
+	hr.mutex.Lock()
+	defer hr.mutex.Unlock()
+	hr.started = true
+}
+
+// CloseSend closes the client connection
+func (hr *bftHeaderReceiver) CloseSend() error {
```

Comment:
should we not wait until `DeliverHeaders()` exits?

Reply:
No, actually `DeliverHeaders` calls this one when it exits. This is used to send a signal to the orderer to stop sending headers and close the connection.